gui/src/ruler.cc

```diff
@@ -0,0 +1,362 @@
+{
+}
+
+Ruler::~Ruler()
+{
+    unregisterDrawingArea();
+}
+
+void Ruler::setDrawingArea(GtkWidget *widget)
+{
+    unregisterDrawingArea();
+    drawingArea = widget;
+
+    width = gtk_widget_get_allocated_width(widget);
+    height = gtk_widget_get_allocated_height(widget);
+
+    // Register callbacks
+    g_signal_connect(widget, "draw", G_CALLBACK(drawCallback), this);
+    g_signal_connect(widget, "size-allocate", G_CALLBACK(sizeAllocateCallback), this);
+}
+
+void Ruler::unregisterDrawingArea()
+{
+    if (drawingArea == nullptr) { return; }
+
+    g_signal_handlers_disconnect_by_data(drawingArea, this);
+    drawingArea = nullptr;
+}
+
+void Ruler::setRange(double lower, double upper)
+{
+    lowerLimit = lower;
+    upperLimit = upper;
+
+    if (drawingArea == nullptr) { return; }
+
+    update();
+
+    // We need to manually trigger the widget to redraw
+    gtk_widget_queue_draw(drawingArea);
+}
+
+void Ruler::update()
+{
+    if (drawingArea == nullptr) { return; }
+
+    // We need to calculate the distance between the largest ticks on the ruler
+    // We will try each interval sequentially until we find an interval which
+    // will produce segments of a large enough width/height when drawn
+
+    // Once we have the distance between the major ticks (the size of the segments)
+    // each segment is then divided into 5 parts assuming there's enough space
+    // and each of those parts is again divided into 2, again, assuming there's enough space
+
+    // Index in the ruler's VALID_INTERVALS array
+    int intervalIndex = 0;
+    // Each interval is multiplied by 10 raised to the power n
+    const int INTERVAL_BASE = 10;
+    int intervalN = 0;
+
+    while (true)
+    {
+        if (VALID_INTERVALS.at(intervalIndex) != 1 || intervalN != 0)
+        {
+            majorInterval = VALID_INTERVALS.at(intervalIndex) * pow(INTERVAL_BASE, intervalN);
+
+            // We check versus the width or height depending on orientation
+            double rulerSize = NAN;
+            if (orientation == HORIZONTAL)
+            {
+                rulerSize = width;
+            }
+            else
+            {
+                rulerSize = height;
+            }
+            // Calculate the drawn size for this interval by mapping from the ruler range
+            // to the ruler size on the screen
+            segmentScreenSize = floor(mapRange(majorInterval, 0.0, upperLimit - lowerLimit, 0.0, rulerSize));
+
+            // If we've found a segment of appropriate size, we can stop
+            if (segmentScreenSize >= MIN_SEGMENT_SIZE) { break; }
+        }
+
+        // Try the next interval
+        intervalIndex++;
+        if (intervalIndex == VALID_INTERVALS.size())
+        {
+            // We tried all intervals for the current n
+            intervalIndex = 0;
+            intervalN++;
+        }
+    }
+}
+
+double Ruler::mapRange(double x, double a_lower, double a_upper, double b_lower, double b_upper)
```

Comment:
"Mapping between ranges" can really mean anything. From the computation, this looks more like a scale operation?

Also, I think the code becomes clearer if you replace `a` and `b` by `from` and `to`